src/cls/async-hooks.ts

```diff
@@ -31,29 +31,63 @@ const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> =
 const WRAPPED = Symbol('@google-cloud/trace-agent:AsyncHooksCLS:WRAPPED');
 
 type ContextWrapped<T> = T&{[WRAPPED]?: boolean};
+type Reference<T> = {
+  value: T
+};
 
 /**
  * An implementation of continuation-local storage on top of the async_hooks
  * module.
  */
 export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
-  private currentContext: {value: Context};
-  private contexts: {[id: number]: Context} = {};
-  private hook: asyncHooksModule.AsyncHook;
+  // instance-scope reference to avoid top-level require.
+  private ah: AsyncHooksModule;
+
+  /** A map of AsyncResource IDs to Context objects. */
+  private contexts: {[id: number]: Reference<Context>} = {};
+  /** The AsyncHook that proactively populates entries in this.contexts. */
+  private hook: asyncHooks.AsyncHook;
+  /** Whether this instance is enabled. */
   private enabled = false;
 
   constructor(private readonly defaultContext: Context) {
-    this.currentContext = {value: this.defaultContext};
-    this.hook = (require('async_hooks') as AsyncHooksModule).createHook({
+    // Store a reference to the async_hooks module, since we will need to query
+    // the current AsyncResource ID often.
+    this.ah = require('async_hooks') as AsyncHooksModule;
+
+    // Create the hook.
+    this.hook = this.ah.createHook({
       init: (id: number, type: string, triggerId: number, resource: {}) => {
-        this.contexts[id] = this.currentContext.value;
-      },
-      before: (id: number) => {
-        if (this.contexts[id]) {
-          this.currentContext.value = this.contexts[id];
+        // init is called when a new AsyncResource is created. We want code
+        // that runs within the scope of this new AsyncResource to see the same
+        // context as its "parent" AsyncResource. The criteria for the parent
+        // depends on the type of the AsyncResource.
+        if (type === 'PROMISE') {
```

Comment:
Can we add a test to make sure that promises end up generating precisely this string as the resource type? If something in node-core changes, let us not break here.

The test can exist either here or in node-core.

Reply:
Done. I think it should exist here, since it'll be hard to keep track of changes to the test elsewhere.